src/utils.js

```diff
@@ -96,49 +99,59 @@ export function get (object, path, fallback) {
 }
 
 /**
- * Immutabily assign a value to a provided object at a given key. If
- * the value is the same, don't do anything. Otherwise return a new
- * object.
+ * Non-destructively assign a value to a provided object at a given key. If the
+ * value is the same, don't do anything. Otherwise return a new object.
  */
-export function set (object, path, value) {
-  if (value === undefined || get(object, path) === value) {
-    return object
-  }
-
+export function set (object, path, goal) {
+  // Ensure we're working with a key path, like: ['a', 'b', 'c']
   path = castPath(path)
 
-  let length = path.length
+  let len  = path.length
 
-  if (length <= 0) {
-    return value
-  } else if (length > 1) {
-    return setIn(object, path, value)
+  if (len <= 0) {
+    return goal
   }
 
-  let copy = clone(object)
-
-  copy[path[0]] = value
-
-  return copy
-}
-
-/**
- * Deeply assign a value given a path of sequential keys.
- */
-export function setIn (object, keys, value) {
-  if (get(object, keys) === value) {
+  if (get(object, path) === goal) {
     return object
   }
 
-  let key = keys[0]
-  let rest = keys.slice(1)
-  let copy = clone(object)
+  let root = clone(object)
+  let node = root
+
+  // For each key in the path...
+  for (var i = 0; i < len; i++) {
+    let key = path[i]
+    let value = goal
+
+    // Are we at the end?
+    if (i < len - 1) {
+      // No: Check to see if the key is already assigned,
+      if (key in node) {
+        // If yes, clone that value
+        value = clone(node[key])
+      } else {
+        // Otherwise assign an object so that we can keep drilling down
+        value = {}
+      }
+    }
 
-  if (rest.length) {
-    copy[key] = (key in copy) ? setIn(copy[key], rest, value) : setIn(Array.isArray(copy) ? [] : {}, rest, value)
-  } else {
-    copy[key] = value
+    // Assign the value, then continue on to the next iteration of the loop
+    // using the next step down
+    node[key] = value
+    node = node[key]
   }
 
-  return copy
+  return root
 }
+
+
+/**
+ * A helper combination of get and set
+ */
+ export function update (state, path, fn, fallback) {
+   let last = get(state, path, fallback)
+   let next = fn(last)
```

Comment:
What if you wanted to use `update` similar to `get` and provide a default value instead of a default function? Will `fn(last)` if it's a `'defaultString'` blow up?